modules/core/src/shaderlib/shadow/shadow.js

```diff
@@ -118,13 +119,40 @@ function _getViewportCenterPosition({viewport, center}) {
   return new Matrix4(viewport.viewProjectionMatrix).invert().transformVector4(center);
 }
 
+function _getViewProjectionMatrix({viewport, shadowMatrix}) {
+  const viewMatrix = shadowMatrix.clone().translate(new Vector3(viewport.center).negate());
+
+  const topLeft = viewport.projectPosition(viewport.unproject([0, 0]));
+  const topRight = viewport.projectPosition(viewport.unproject([viewport.width, 0]));
+  const bottomLeft = viewport.projectPosition(viewport.unproject([0, viewport.height]));
+  const bottomRight = viewport.projectPosition(
+    viewport.unproject([viewport.width, viewport.height])
+  );
+
+  const pos0 = viewMatrix.transformVector3([topLeft[0], topLeft[1], 0]);
+  const pos1 = viewMatrix.transformVector3([topRight[0], topRight[1], 0]);
+  const pos2 = viewMatrix.transformVector3([bottomLeft[0], bottomLeft[1], 0]);
+  const pos3 = viewMatrix.transformVector3([bottomRight[0], bottomRight[1], 0]);
+
+  const projectionMatrix = new Matrix4().ortho({
+    left: Math.min(pos0[0], pos1[0], pos2[0], pos3[0]),
+    right: Math.max(pos0[0], pos1[0], pos2[0], pos3[0]),
+    bottom: Math.min(pos0[1], pos1[1], pos2[1], pos3[1]),
+    top: Math.max(pos0[1], pos1[1], pos2[1], pos3[1]),
+    // Near plane could be too close to cover tall objects, scale up by 2.0
+    near: Math.min(-pos0[2], -pos1[2], -pos2[2], -pos3[2]) * 2.0,
```

Comment:
Does this mean the shadow map is not covering the closest corner?

Reply:
it covers the closest ground corner but may not cover the closest 3d object top.